### PR TITLE
HIVE-25032: Optimise PartitionManagementTask

### PR DESCRIPTION
**Pull request by dengzhhu653**
### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
1. the added unit test
2. benchmark 

```shell
  java -Xmx8g -jar hmsbench-jar-with-dependencies.jar --confdir hiveconf --runMode MSCK -d bm -N 1000 -L 10 -W 2 -H localhost -P 9083
```
The above command will load the configuration in hiveconf directory, run the benchmark 10 times, each time there are 200 in 1000 tables needed to recover partitions, and the 200 tables each have 500 partitions to add and 500 partitions to drop. 
In hiveconf directory, only core-site.xml is here for specifying local FileSystem.
- After optimization:
Operation                      Mean     Med      Min      Max      Err%
~~PartitionManagementTask        1469     1437     1394     1587     4.744~~
~~PartitionManagementTask.1000   1.431e+05 1.406e+05 1.382e+05 1.555e+05 3.959~~
Operation                      Mean     Med      Min      Max      Err%
PartitionManagementTask        1867     1853     1644     2330     11.33
PartitionManagementTask.1000   1.637e+05 1.655e+05 1.494e+05 1.691e+05 3.493
- Before optimization:
Operation                      Mean     Med      Min      Max      Err%
PartitionManagementTask        7443     7438     7371     7500     0.5184
PartitionManagementTask.1000   7.266e+05 7.271e+05 7.168e+05 7.359e+05 0.7314

### CODE REVIEW

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java [hunk 1]
```diff
@@ -2174,6 +2174,22 @@ List<Partition> dropPartitions(String catName, String dbName, String tblName,
...
+   * @param tblName Name of the table
+   * @param partNames Arrays of partitions to drop
+   * @param options Boolean options for dropping partitions
+   * @return List of Partitions dropped
+   * @throws NoSuchObjectException No partition matches the expression(s), and ifExists was false.
+   * @throws MetaException error access the RDBMS or storage.
+   * @throws TException On failure
+   */
+  List<Partition> dropPartitions(String catName, String dbName, String tblName,
+                                 String[] partNames,
```
Comment (saihemanth-cloudera) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java:2189]: Why do we want to declare this as String array? We are anyway using List<String> everywhere, right?

**Review by saihemanth-cloudera — COMMENTED** [review]
Minor comments. Otherwise looks good.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java (hunk 1 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java]: Thanks @saihemanth-cloudera for the comment. Introduce `List<Partition> dropPartitions(DropPartitionsRequest req, PartitionDropOptions options)` instead, support HMS deleting partitions in batch by the provided partition names.

**Review by saihemanth-cloudera — COMMENTED** [review]
I think the test failures are genuine. Can you please address them?

**Comment by dengzhhu653** [comment]
> I think the test failures are genuine. Can you please address them?

ack

**Comment by dengzhhu653** [comment]
> LGTM +1

Thanks @saihemanth-clouder. Please hold on the merge, I'm thinking how not to introduce the new `dropPartitions` in IMetaStoreClient`.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PartitionIterable.java [hunk 2]
```diff
@@ -103,11 +107,17 @@ private void getNextBatch() {
           batch_counter++;
         }
         try {
-          String dbName = prependCatalogToDbName(table.getCatName(), table.getDbName(), null);
-          GetPartitionsByNamesRequest req =
-              convertToGetPartitionsByNamesRequest(dbName, table.getTableName(), nameBatch);
-          batchIter =
-            msc.getPartitionsByNames(req).getPartitionsIterator();
+          if (request != null) {
+            request.setFilterSpec(new GetPartitionsFilterSpec());
```
Comment (henrib) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PartitionIterable.java:111]: May be introduce a local variable for 'new GetPartitionsFilterSpec' and avoid the getters after.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/Msck.java [hunk 3]
```diff
@@ -589,17 +593,22 @@ public Void execute(int size) throws MetastoreException {
       }
 
       private List<Pair<Integer, byte[]>> getPartitionExpr(final List<String> parts) throws MetaException {
-        List<Pair<Integer, byte[]>> expr = new ArrayList<>(parts.size());
+        StringBuilder exprBuilder = new StringBuilder();
+        String orExpr = " OR ";
         for (int i = 0; i < parts.size(); i++) {
```
[the remark this answers is not in the document]
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/Msck.java:598]: Sure, will change that. Thank you for your comments.

**Pull request merged**